python/lib/core/dmod/core/common/collection.py

```diff
@@ -235,6 +238,165 @@ def __contains__(self, obj: object) -> bool:
         return obj in self.__data
 
 
+class _OccurrenceTracker(typing.Generic[_T]):
+    """
+    Keeps track of occurrences of a type of value that have been encountered within a duration
+    """
+    def __init__(self, key: _T, duration: timedelta, threshold: int, on_filled: typing.Callable[[_T], typing.Any]):
+        self.__key = key
+        self.__duration = duration
+        self.__threshold = threshold
+        self.__occurences: typing.List[datetime] = []
+        self.__on_filled = on_filled
+
+    def value_encountered(self):
+        """
+        Inform the tracker that the value has been encountered again
+        """
+        self.update_occurrences()
+        self.__occurences.append(datetime.now())
+        if len(self.__occurences) >= self.__threshold:
+            self.__on_filled(self.__key)
+
+    def update_occurrences(self) -> int:
+        """
+        Update the list of occurrences to include only those within the current duration
+
+        Returns:
+            The number of occurrences still being tracked
+        """
+        cutoff: datetime = datetime.now() - self.__duration
+        self.__occurences = sorted([
```

Comment:
I'm investigating.

Reply:
I don't think `heapq` would be appropriate here, partially because I'm not even sure that _ordering_ is appropriate here. It's only supposed to be a list of list of times that something occurred that is routinely filtered and sized. `heapq` doesn't have any filtering ops, order doesn't particularly matter (and would probably _always_ be in order anyways), and `heapq` doesn't offer much if any particular insight on better ways to count.

I'm just going to remove the sorting operation.